---
title: "3-composition_overview.Rmd"
author: "Antton Alberdi"
date: "2023-12-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries

```{r libraries, warning=FALSE, comments="", message=FALSE}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ape))
suppressPackageStartupMessages(library(distillR))
suppressPackageStartupMessages(library(ggtree))
suppressPackageStartupMessages(library(ggnewscale))
suppressPackageStartupMessages(library(ggtreeExtra))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(phytools))
suppressPackageStartupMessages(library(Rtsne))
suppressPackageStartupMessages(library(ggh4x))
suppressPackageStartupMessages(library(hilldiv2))
```

## Load data

```{r load_data}
load("data/data.Rdata")
```

## Sort samples

```{r sort_samples}
#Arranged by days
samples_days <- sample_metadata %>%
  arrange(day,treatment)
```

## Taxonomy barplot

```{r barplot_days, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>%
    unique() %>%
    arrange(phylum) %>%
    select(colors) %>%
    pull()

genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(., sample_metadata, by = join_by(sample == sample)) %>% #append sample metadata
  filter(treatment != "TG0") %>%
  ggplot(., aes(x=sample,y=count, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(y = "Relative abundance") +
    facet_nested(. ~ day + treatment,  scales="free_x") + #facet per day and treatment
    guides(fill = guide_legend(ncol = 1)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black")) +
   labs(fill="Phylum")

```

```{r barplot_treatment, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(., sample_metadata, by = join_by(sample == sample)) %>% #append sample metadata
  mutate(day = factor(day, levels = sort(unique(sample_metadata$day)))) %>% #sort animals per treatment
  filter(treatment != "TG0") %>%
  ggplot(., aes(x=sample,y=count, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(y = "Relative abundance") +
    facet_nested(. ~ treatment + day,  scales="free_x") + #facet per treatment and animal
    guides(fill = guide_legend(ncol = 1)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black")) +
   labs(fill="Phylum")

```

## Genome count table (digesta)
```{r genomecounts_faeces, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
vertical_tree <- force.ultrametric(genome_tree,method="extend") %>%
        ggtree(., size = 0.3)

#Add phylum colors
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    mutate(phylum = factor(phylum, levels = unique(phylum))) %>%
    column_to_rownames(var = "genome") %>%
    select(phylum)


colors_alphabetic <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
  select(phylum, colors) %>%
    unique() %>%
    arrange(phylum) %>%
    select(colors) %>%
    pull()

vertical_tree <- gheatmap(vertical_tree, phylum_colors, offset=-0.6, width=0.1, colnames=FALSE) +
    scale_fill_manual(values=colors_alphabetic) +
    new_scale_fill()

#Add genome counts of d0
genome_counts_faeces_d0_d0 <- genome_counts %>%
          select(all_of(c("genome",sample_metadata %>% filter(type=="faeces") %>% filter(day=="0") %>% select(sample) %>% pull()))) %>% 
          column_to_rownames(var="genome") %>% tss()

vertical_tree <- gheatmap(vertical_tree, log10(genome_counts_faeces_d0_d0), offset=-0.4, width=0.3, colnames=FALSE, colnames_angle=90, font.size=3, colnames_position="top", colnames_offset_y = 15) +
    vexpand(.08) +
    coord_cartesian(clip = "off") +
    scale_fill_gradient(low = "lightblue", high = "#315b7d", na.value="#f4f4f4") +
    new_scale_fill()

#Add mean values of d0
genome_counts_faeces_d0_mean <- genome_counts_digesta_d0 %>% 
            rownames_to_column(var="genome") %>% 
            rowwise() %>% 
            mutate(mean = mean(c_across(where(is.numeric)))) %>% 
            select(genome,mean)

vertical_tree <-  vertical_tree +
        scale_fill_manual(values = "#cccccc") +
        geom_fruit(
             data=genome_counts_digesta_d0_mean,
             geom=geom_bar,
             mapping = aes(x=mean, y=genome),
             pwidth = 0.1,
             offset = 0.15,
             width= 1,
             orientation="y",
             axis.params=list(axis="x"),
             stat="identity") +
        new_scale_fill()

#Add genome counts of d7
genome_counts_faeces_d7 <- genome_counts %>%
          select(all_of(c("genome",sample_metadata %>% filter(type=="faeces") %>% filter(day=="7") %>% select(sample) %>% pull()))) %>% 
          column_to_rownames(var="genome") %>% tss()

vertical_tree <- gheatmap(vertical_tree, log10(genome_counts_faeces_d7), offset=0.6, width=0.3, colnames=FALSE, colnames_angle=90, font.size=3, colnames_position="top", colnames_offset_y = 15) +
    vexpand(.08) +
    coord_cartesian(clip = "off") +
    scale_fill_gradient(low = "lightblue", high = "#315b7d", na.value="#f4f4f4") +
    new_scale_fill()


#Add mean values of d7
genome_counts_faeces_d7_mean <- genome_counts_faeces_d7 %>% 
            rownames_to_column(var="genome") %>% 
            rowwise() %>% 
            mutate(mean = mean(c_across(where(is.numeric)))) %>% 
            select(genome,mean)

vertical_tree <-  vertical_tree +
        scale_fill_manual(values = "#cccccc") +
        geom_fruit(
             data=genome_counts_faeces_d7_mean,
             geom=geom_bar,
             mapping = aes(x=mean, y=genome),
             pwidth = 0.1,
             offset = 0.3,
             width= 1,
             orientation="y",
             axis.params=list(axis="x"),
             stat="identity") +
        new_scale_fill()

#Add genome counts of d14
genome_counts_faeces_d14 <- genome_counts %>%
          select(all_of(c("genome",sample_metadata %>% filter(type=="faeces") %>% filter(day=="14") %>% select(sample) %>% pull()))) %>% 
          column_to_rownames(var="genome") %>% tss()

vertical_tree <- gheatmap(vertical_tree, log10(genome_counts_faeces_d14), offset=1.7, width=0.3, colnames=FALSE, colnames_angle=90, font.size=3, colnames_position="top", colnames_offset_y = 15) +
    vexpand(.08) +
    coord_cartesian(clip = "off") +
    scale_fill_gradient(low = "lightblue", high = "#315b7d", na.value="#f4f4f4") +
    new_scale_fill()


#Add mean values of d14
genome_counts_faeces_d14_mean <- genome_counts_faeces_d14 %>% 
            rownames_to_column(var="genome") %>% 
            rowwise() %>% 
            mutate(mean = mean(c_across(where(is.numeric)))) %>% 
            select(genome,mean)

vertical_tree <-  vertical_tree +
        scale_fill_manual(values = "#cccccc") +
        geom_fruit(
             data=genome_counts_faeces_d14_mean,
             geom=geom_bar,
             mapping = aes(x=mean, y=genome),
             pwidth = 0.1,
             offset = 0.33,
             width= 1,
             orientation="y",
             axis.params=list(axis="x"),
             stat="identity") +
        new_scale_fill()

#Add genome counts of d21
genome_counts_faeces_d21 <- genome_counts %>%
          select(all_of(c("genome",sample_metadata %>% filter(type=="faeces") %>% filter(day=="21") %>% select(sample) %>% pull()))) %>% 
          column_to_rownames(var="genome") %>% tss()

vertical_tree <- gheatmap(vertical_tree, log10(genome_counts_faeces_d21), offset=2.7, width=0.3, colnames=FALSE, colnames_angle=90, font.size=3, colnames_position="top", colnames_offset_y = 15) +
    vexpand(.08) +
    coord_cartesian(clip = "off") +
    scale_fill_gradient(low = "lightblue", high = "#315b7d", na.value="#f4f4f4") +
    new_scale_fill()


#Add mean values of d21
genome_counts_faeces_d21_mean <- genome_counts_faeces_d21 %>% 
            rownames_to_column(var="genome") %>% 
            rowwise() %>% 
            mutate(mean = mean(c_across(where(is.numeric)))) %>% 
            select(genome,mean)

vertical_tree <-  vertical_tree +
        scale_fill_manual(values = "#cccccc") +
        geom_fruit(
             data=genome_counts_faeces_d21_mean,
             geom=geom_bar,
             mapping = aes(x=mean, y=genome),
             pwidth = 0.1,
             offset = 0.3,
             width= 1,
             orientation="y",
             axis.params=list(axis="x"),
             stat="identity") +
        new_scale_fill()

vertical_tree +
  theme(legend.position='none')
```

### Top genera per treatment/time

```{r top_genera_rank}
genus_rank <- genome_counts %>%
    pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
    left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append taxonomy
    inner_join(., sample_metadata, by = join_by(sample == sample)) %>% #append metadata
    group_by(genus) %>%
    summarise(count=sum(count)) %>%
    arrange(-count) %>%
    select(genus) %>%
    slice(1:30) %>%
    pull()
```

```{r top_genera_plot, message=FALSE, warning=FALSE, fig.height=10, fig.width=10, fig.fullwidth=TRUE}
genome_counts %>%
    pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
    left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append taxonomy
    inner_join(., sample_metadata, by = join_by(sample == sample)) %>% #append metadata
    group_by(sample,treatment,day,genus) %>%
    summarise(count=sum(count)) %>%
    filter(genus %in% genus_rank) %>%
    mutate(genus = fct_relevel(genus, rev(genus_rank))) %>%
    ggplot(., aes(y=genus,x=count)) +
      geom_col() +
      facet_nested(. ~ day + treatment) +
      theme(axis.text.x=element_blank(), #remove x axis labels
        axis.ticks.x=element_blank(), #remove x axis ticks
        )
      labs(y="Top 30 genera",x="Genome counts")
```
## Alpha diversities
```{r alpha_div, message=FALSE, warning=FALSE, fig.height=10, fig.width=10, fig.fullwidth=TRUE}

#Calculate Hill numbers
richness <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=0) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(richness=1) %>%
            rownames_to_column(var="sample")

neutral <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(neutral=1) %>%
            rownames_to_column(var="sample")

phylogenetic <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1,tree=genome_tree) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(phylogenetic=1) %>%
            rownames_to_column(var="sample")

# Aggregate basal GIFT into elements
dist <- genome_gifts %>%
    to.elements(., GIFT_db) %>%
    traits2dist(., method="gower")

functional <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1,dist=dist) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(functional=1) %>%
            rownames_to_column(var="sample") %>%
            mutate(functional = if_else(is.nan(functional), 1, functional))



#Merge into a single table
#alpha_div <- cbind(sample=colnames(genome_counts[-1]),richness=q0n,neutral=round(q1n,3),phylo=round(q1p,3),func=round(q1f,3)) %>%
alpha_div <- richness %>%
      full_join(neutral,by=join_by(sample==sample)) %>%
      full_join(phylogenetic,by=join_by(sample==sample)) %>%
      full_join(functional,by=join_by(sample==sample)) %>%
      pivot_longer(-sample, names_to = "data", values_to = "value") %>%
      left_join(., sample_metadata, by = join_by(sample == sample)) %>%
      mutate(value = as.numeric(value)) %>%
      mutate(data = factor(data, levels = c("richness","neutral","phylogenetic","functional"))) #sort metrics


#Plot diversities
alpha_plot_neutral <- alpha_div %>%
        filter(data=="neutral") %>%
        ggplot(aes(x=value, y=sample)) +
            geom_bar(stat='identity', fill="#6c9ebc") +
            facet_nested(day + treatment ~ data,  scales="free") + #facet per treatment and animal
            coord_cartesian(xlim = c(1, NA)) +
            theme_classic() +
            theme(
                strip.background = element_blank(),
                panel.grid.minor.x = element_line( size=.1, color="grey" ),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                axis.text.x = element_text(angle = 45, hjust = 1),
                strip.text.y = element_blank()
            )

alpha_plot_phylo <- alpha_div %>%
        filter(data=="phylogenetic") %>%
        ggplot(aes(x=value, y=sample)) +
            geom_bar(stat='identity', fill="#6c9ebc") +
            facet_nested(day + treatment ~ data,  scales="free") + #facet per treatment and animal
            coord_cartesian(xlim = c(1, NA)) +
            theme_classic() +
            theme(
                strip.background = element_blank(),
                panel.grid.minor.x = element_line( size=.1, color="grey" ),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                axis.text.y = element_blank(),
                axis.text.x = element_text(angle = 45, hjust = 1),
                strip.text.y = element_blank()
            )

alpha_plot_func <- alpha_div %>%
        filter(data=="functional") %>%
        ggplot(aes(x=value, y=sample)) +
            geom_bar(stat='identity', fill="#6c9ebc") +
            facet_nested(day + treatment ~ data,  scales="free") + #facet per treatment and animal
            coord_cartesian(xlim = c(1, NA)) +
            theme_classic() +
            theme(
                strip.background = element_blank(),
                #panel.grid.minor.x = element_line( size=.1, color="grey" ),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                axis.text.y = element_blank(),
                axis.text.x = element_text(angle = 45, hjust = 1)
            )

grid.arrange(alpha_plot_neutral, alpha_plot_phylo, alpha_plot_func, nrow = 1)

```
