---
title: "5-hmsc.Rmd"
author: "Antton Alberdi"
date: "2024-01-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries

```{r libraries, warning=FALSE, comments="", message=FALSE}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(phyloseq))
suppressPackageStartupMessages(library(Hmsc))
suppressPackageStartupMessages(library(distillR))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(abind))

```

## Load data

```{r load_data}
load("data/data.Rdata")
```

## Prepare input for Hmsc

Only using faecal samples.

```{r hmsc_input, warning=FALSE, comments="", message=FALSE}
# Random effects data (study design)
StudyDesign <- sample_metadata %>% 
                    mutate(sample2=sample) %>% 
                    column_to_rownames("sample2") %>% 
                    mutate(sample = factor(sample)) %>% 
                    filter(treatment != "TG0") %>%  #remove controls
                    select(sample)

# Genome count table (quantitative community data)
YData <- read_counts %>% 
                    mutate(across(where(is.numeric), ~ . +1 )) %>% #add +1 pseudocount to remove zeros
                    mutate(across(where(is.numeric), ~ . / (genome_metadata$length / 150) )) %>% #transform to genome counts
                    mutate(across(where(is.numeric), ~  log(.) )) %>% #log-transform
                    arrange(genome) %>%
                    column_to_rownames("genome") %>% 
                    select(all_of(row.names(StudyDesign))) %>%  #filter only faecal samples
                    as.data.frame() %>%
                    t() # transpose

# Fixed effects data (explanatory variables)
XData <- sample_metadata %>% 
                    column_to_rownames("sample") %>% 
                    filter(treatment != "TG0") %>% #remove controls
                    mutate(logseqdepth=read_counts %>% #total log-sequencing depth
                        select(all_of(row.names(StudyDesign))) %>% 
                        colSums() %>% 
                        log()
                    ) %>% 
                    mutate(treatment = factor(treatment, levels = c("TG1","TG2","TG3","TG4","TG5"))) %>% 
                    mutate(day=as.numeric(day)) %>% 
                    mutate(logseqdepth=as.numeric(logseqdepth)) %>% 
                    select(day, treatment, logseqdepth)

# Genome trait data
TrData <- genome_gifts %>% 
                    to.elements(.,GIFT_db) %>% 
                    to.functions(.,GIFT_db) %>% 
                    as.data.frame() %>%
                    rownames_to_column(var = "genome") %>%
                    arrange(genome) %>%
                    column_to_rownames(var="genome")

# Genome phylogeny
PData <- genome_tree
```

## Define formulas of the Hmsc model

```{r hmsc_formulas, warning=FALSE, comments="", message=FALSE}

# Fixed effects formula
XFormula = ~day*treatment + logseqdepth

# Trait formula
TrFormula = ~B01+B02+B03+B04+B07+D01+D02+D03+D05+D06+D07

# Study design
rL.sample = HmscRandomLevel(units = levels(StudyDesign$sample))
```

## Define and Hmsc models
```{r hmsc_models, warning=FALSE, comments="", message=FALSE}
#Define models
model1 = Hmsc(Y=YData,
         XData = XData, 
         XFormula = XFormula,
         studyDesign = StudyDesign,
         phyloTree = PData, 
         ranLevels = list("sample"=rL.sample),
         TrData = TrData, 
         TrFormula = TrFormula, 
         distr = "normal",
         YScale = TRUE)

#Save list of models as an R object.
model_list = list(model1=model1)
if (!dir.exists("hmsc")){dir.create("hmsc")}
save(model_list, file = "hmsc/hmsc.Rdata")
```

Upload **hmsc/hmsc.Rdata** to the HPC respecting the directory structure.

## Define MCMC
```{r hmsc_mcmc, warning=FALSE, comments="", message=FALSE}
# How often to sample the MCMC
MCMC_samples_list = 250

# The number of MCMC steps between each recording sample
MCMC_thin_list = c(1, 10)

# The number of MCMC chains to use
nChains = 4
```

## Generate Hmsc executables

The next chunk generates shell files for every combination of model, MCMC samples and MCMM thinning, ready to be launched as SLURM jobs.

```{r hmsc_executables, warning=FALSE, comments="", message=FALSE}

modelchains <- expand.grid(model = names(model_list), sample = MCMC_samples_list, thin = MCMC_thin_list)

if (!dir.exists("hmsc")){dir.create("hmsc")}
for(i in c(1:nrow(modelchains))){
      modelname=as.character(modelchains[i,1])
      sample=modelchains[i,2]
      thin=modelchains[i,3]
      executablename <- paste0("hmsc/exe_",modelname,"_",sample,"_",thin,".sh")
      fitname <- paste0("hmsc/fit_",modelname,"_",sample,"_",thin,".Rdata")
      convname <- paste0("hmsc/conv_",modelname,"_",sample,"_",thin,".Rdata")
      model <- paste0('model_list$',modelname)
      psrf.beta.name <-  paste0("psrf.beta.",modelname,"_",sample,"_",thin)
      psrf.gamma.name <-  paste0("psrf.gamma.",modelname,"_",sample,"_",thin)
      psrf.rho.name <-  paste0("psrf.rho.",modelname,"_",sample,"_",thin)
      jobname <- paste0("hmsc_",modelname,"_",sample,"_",thin)
      minutes <- round(sample * thin / 4, 0) 
      code <- sprintf("#!/bin/bash
#SBATCH --job-name=%s                   # Job name
#SBATCH --nodes=1
#SBATCH --ntasks=4                      # Run on 4 CPUs
#SBATCH --mail-user=antton.alberdi@sund.ku.dk
#SBATCH --mem=96gb                      # Job memory request
#SBATCH --time=%d                       # In minutes

# Create conda environment if needed
module load mamba/1.3.1
if ! conda info --envs | grep -q hmsc; then
  conda create -p ./hmsc/hmsc_env -y r-essentials r-base r-tidyverse r-Hmsc
fi

# Activate conda environment
source activate ./hmsc/hmsc_env

# Run R script
Rscript -e '
library(tidyverse)
library(Hmsc)
# Load formulas and data
load(\"hmsc/hmsc.Rdata\")

# Declare placeholders
modelname = \"%s\"
model = %s
fitname = \"%s\"
convname = \"%s\"
sample = %d
thin = %d
nchains = %d

# Run model fitting
m = sampleMcmc(hM = model, 
         samples = sample, 
         thin = thin,
         adaptNf=rep(ceiling(0.4*sample*thin),model$nr),
         transient = ceiling(0.5*sample*thin),
         nChains = nchains,
         nParallel = nchains)
         
# Assess chain convergence
mpost = convertToCodaObject(m, 
      spNamesNumbers = c(T,F), 
      covNamesNumbers = c(T,F),
      Beta = TRUE,
      Gamma = TRUE,
      V = FALSE,
      Sigma = FALSE,
      Rho = TRUE,
      Eta = FALSE,
      Lambda = FALSE,
      Alpha = FALSE,
      Omega = FALSE,
      Psi = FALSE,
      Delta = FALSE) # Convert to CODA object

# Fixed effects
assign(paste0(\"psrf.beta.\", modelname,\"_\",sample,\"_\",thin), gelman.diag(mpost$Beta,multivariate=FALSE)$psrf)

# Traits
assign(paste0(\"psrf.gamma.\", modelname,\"_\",sample,\"_\",thin), gelman.diag(mpost$Gamma,multivariate=FALSE)$psrf)

# Phylogeny
assign(paste0(\"psrf.rho.\", modelname,\"_\",sample,\"_\",thin), gelman.diag(mpost$Rho,multivariate=FALSE)$psrf)

# Write convergence data
save(%s, %s, %s, file=convname)

# Save model fit object
save(m, file=fitname)
'
", jobname, minutes, modelname, model, fitname, convname, sample, thin, nChains, psrf.beta.name, psrf.gamma.name, psrf.rho.name)
      writeLines(code, executablename)
    }
```

Upload the produced **hmsc/exe_XXXXX.sh** files to the HPC respecting the directory structure.

## Fit Hmsc models (in Mjolnir HPC)

Launch the SLURM jobs by using:

```{sh, eval=FALSE}
#Create and define tmpdir
tmpdir="./tmp"
mkdir -p "$tmpdir"
export TMPDIR="$tmpdir"

# Submit all .sh files in the hmsc folder
for jobfile in hmsc/exe_*.sh; do
    sbatch "$jobfile"
done

#Or launch them one by one only the ones you want to launch
sbatch hmsc/exe_model1_250_1.sh
sbatch hmsc/exe_model1_250_10.sh
sbatch hmsc/exe_model1_250_100.sh
sbatch hmsc/exe_model1_250_1000.sh
```

## Assess chaing convergence

Convergence diagnostic values substantially above 1 indicate lack of convergence.
Values below 1.1 are considered good enough

```{r hmsc_convergence, warning=FALSE, comments="", message=FALSE}

# Load all conv file available in the hmsc folder
list.files(path = "hmsc", pattern = "^conv_", full.names = TRUE, include.dirs = TRUE) %>%
  lapply(.,load,.GlobalEnv)

# Create a merged psrf.beta (genome) plot
ls() %>% 
        grep("^psrf\\.beta", ., value = TRUE) %>% 
        map_dfr(~ {
         mat <- get(.x)
          data.frame(modelchain = .x, as.data.frame(mat, , stringsAsFactors = FALSE)) %>% 
              rownames_to_column(var="parameter") %>%
              mutate(model = str_split(modelchain, "_") %>% map_chr(1) %>% gsub("psrf.beta.","",.)) %>%
              mutate(sample = str_split(modelchain, "_") %>% map_chr(2)) %>% #extract sample info from model name
              mutate(thin = str_split(modelchain, "_") %>% map_chr(3)) #extract thin info from model name
      }) %>% 
      ggplot(.,aes(x=reorder(modelchain,-Point.est.,fun=function(x) {quantile(x, probs = 0.9)}),y=Point.est.)) +
        geom_violin(fill="#b8d9e3", color="#328da8") +
        geom_jitter(alpha=0.3,size=0.2, color="#a8babf") +
        stat_summary(fun=function(x) {quantile(x, probs = 0.9)}, geom="crossbar", width=0.2, color="orange") +
        geom_hline(yintercept=1.1, linetype="dashed", color = "red") +
        ylim(0.9,2)+
        labs(x="Model chains",y="Parameter estimates")+
        theme_classic()
      
# Create a merged psrf.gamma (trait) plot
ls() %>% 
        grep("^psrf\\.gamma", ., value = TRUE) %>% 
        map_dfr(~ {
         mat <- get(.x)
          data.frame(modelchain = .x, as.data.frame(mat, , stringsAsFactors = FALSE)) %>% 
              rownames_to_column(var="parameter") %>%
              mutate(model = str_split(modelchain, "_") %>% map_chr(1) %>% gsub("psrf.gamma.","",.)) %>%
              mutate(sample = str_split(modelchain, "_") %>% map_chr(2)) %>% #extract sample info from model name
              mutate(thin = str_split(modelchain, "_") %>% map_chr(3)) #extract thin info from model name
      }) %>% 
      ggplot(.,aes(x=reorder(modelchain,-Point.est.,fun=function(x) {quantile(x, probs = 0.9)}),y=Point.est.)) +
        geom_violin(fill="#b8d9e3", color="#328da8") +
        geom_jitter(alpha=0.3,size=0.2, color="#a8babf") +
        stat_summary(fun=function(x) {quantile(x, probs = 0.9)}, geom="crossbar", width=0.2, color="orange") +
        geom_hline(yintercept=1.1, linetype="dashed", color = "red") +
        ylim(0.9,2)+
        labs(x="Model chains",y="Parameter estimates")+
        theme_classic()
      

# Create a merged psrf.rho (phylogeny) plot
ls() %>% 
        grep("^psrf\\.rho", ., value = TRUE) %>% 
        map_dfr(~ {
         mat <- get(.x)
          data.frame(modelchain = .x, as.data.frame(mat, , stringsAsFactors = FALSE)) %>% 
              rownames_to_column(var="parameter") %>%
              mutate(model = str_split(modelchain, "_") %>% map_chr(1) %>% gsub("psrf.beta.","",.)) %>%
              mutate(sample = str_split(modelchain, "_") %>% map_chr(2)) %>% #extract sample info from model name
              mutate(thin = str_split(modelchain, "_") %>% map_chr(3)) #extract thin info from model name
      }) %>% 
      ggplot(.,aes(x=reorder(modelchain,-Point.est.,fun=function(x) {quantile(x, probs = 0.9)}),y=Point.est.)) +
        geom_violin(fill="#b8d9e3", color="#328da8") +
        geom_jitter(alpha=0.3,size=0.2, color="#a8babf") +
        stat_summary(fun=function(x) {quantile(x, probs = 0.9)}, geom="crossbar", width=0.2, color="orange") +
        geom_hline(yintercept=1.1, linetype="dashed", color = "red") +
        ylim(0.9,2)+
        labs(x="Model chains",y="Parameter estimates")+
        theme_classic()



```

## Compute variance partitioning

```{r hmsc_variancepart, warning=FALSE, comments="", message=FALSE}

# Select modelchain of interest
load("hmsc/fit_model1_250_1.Rdata")

varpart=computeVariancePartitioning(m)
plotVariancePartitioning(hM=m,VP=varpart)
```

## Predict responses

```{r hmsc_temporal_predictions, warning=FALSE, comments="", message=FALSE}

# Select modelchain of interest
load("hmsc/fit_model1_250_1.Rdata")

gradient = c(7:35)
gradientlength = length(gradient)

#Treatment-specific gradient predictions
pred_TG1 <- constructGradient(m, 
                      focalVariable = "day", 
                      non.focalVariables = list(logseqdepth=list(1),treatment=list(3,"TG1")), 
                      ngrid=gradientlength) %>%
             predict(m, Gradient = ., expected = TRUE)

pred_TG2 <- constructGradient(m, 
                      focalVariable = "day", 
                      non.focalVariables = list(logseqdepth=list(1),treatment=list(3,"TG2")), 
                      ngrid=gradientlength) %>%
             predict(m, Gradient = ., expected = TRUE)

pred_TG3 <- constructGradient(m, 
                      focalVariable = "day", 
                      non.focalVariables = list(logseqdepth=list(1),treatment=list(3,"TG3")), 
                      ngrid=gradientlength) %>%
             predict(m, Gradient = ., expected = TRUE)

pred_TG4 <- constructGradient(m, 
                      focalVariable = "day", 
                      non.focalVariables = list(logseqdepth=list(1),treatment=list(3,"TG4")), 
                      ngrid=gradientlength) %>%
             predict(m, Gradient = ., expected = TRUE)

pred_TG5 <- constructGradient(m, 
                      focalVariable = "day", 
                      non.focalVariables = list(logseqdepth=list(1),treatment=list(3,"TG5")), 
                      ngrid=gradientlength) %>%
             predict(m, Gradient = ., expected = TRUE)
```

```{r hmsc_temporal_predictions_genomes, warning=FALSE, comments="", message=FALSE}

#Treatment-specific genome predictions
predY_TG1 <- pred_TG1 %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG1",gradientlength*1000)) %>%
        pivot_longer(-c(day,treatment), names_to = "genome", values_to = "value")

predY_TG2 <- pred_TG2 %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG2",gradientlength*1000)) %>%
        pivot_longer(-c(day,treatment), names_to = "genome", values_to = "value")

predY_TG3 <- pred_TG3 %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG3",gradientlength*1000)) %>%
        pivot_longer(-c(day,treatment), names_to = "genome", values_to = "value")

predY_TG4 <- pred_TG4 %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG4",gradientlength*1000)) %>%
        pivot_longer(-c(day,treatment), names_to = "genome", values_to = "value")

predY_TG5 <- pred_TG5 %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG5",gradientlength*1000)) %>%
        pivot_longer(-c(day,treatment), names_to = "genome", values_to = "value")

predY <- rbind(predY_TG1,predY_TG2,predY_TG3,predY_TG4,predY_TG5)

```

```{r hmsc_temporal_predictions_genomes_compare}
predY_summ <- predY %>%
  filter(!is.infinite(value) | is.na(value)) %>%
  group_by(genome, day, treatment) %>%
  summarize(mean = mean(value, na.rm = TRUE),
            q0.1 = quantile(value, 0.05, na.rm = TRUE),
            q0.9 = quantile(value, 0.95, na.rm = TRUE))

# Overall trends
predY_asc <- predY_summ %>%
    pivot_wider(names_from = day, values_from = c(mean,q0.1,q0.9))  %>%
    filter(q0.9_7 < q0.1_35) %>%
    select(genome) %>%
    unique() %>% pull()

predY_desc <- predY_summ %>%
    pivot_wider(names_from = day, values_from = c(mean,q0.1,q0.9))  %>%
    filter(q0.1_7 > q0.9_35) %>%
    select(genome) %>%
    unique() %>% pull()

# Across treatment differences 
predY_dynamics <- predY_summ %>%
    pivot_wider(names_from = c(day,treatment), values_from = c(mean,q0.1,q0.9))  %>%
    mutate(TG1 = ifelse(q0.9_7_TG1 < q0.1_35_TG1, "asc", "inv"))  %>%
    mutate(TG1 = ifelse(q0.1_7_TG1 > q0.9_35_TG1, "dec", "inv"))  %>%
    mutate(TG2 = ifelse(q0.9_7_TG2 < q0.1_35_TG2, "asc", "inv"))  %>%
    mutate(TG2 = ifelse(q0.1_7_TG2 > q0.9_35_TG2, "dec", "inv"))  %>%
    mutate(TG3 = ifelse(q0.9_7_TG3 < q0.1_35_TG3, "asc", "inv"))  %>%
    mutate(TG3 = ifelse(q0.1_7_TG3 > q0.9_35_TG3, "dec", "inv"))  %>%
    mutate(TG4 = ifelse(q0.9_7_TG4 < q0.1_35_TG4, "asc", "inv"))  %>%
    mutate(TG4 = ifelse(q0.1_7_TG4 > q0.9_35_TG4, "dec", "inv"))  %>%
    mutate(TG5 = ifelse(q0.9_7_TG5 < q0.1_35_TG5, "asc", "inv"))  %>%
    mutate(TG5 = ifelse(q0.1_7_TG5 > q0.9_35_TG5, "dec", "inv"))  %>%
    mutate(dynamics = ifelse(any(!grepl("inv", c(TG1, TG2, TG3, TG4, TG5))), "variable", "invariable")) %>%
    filter(dynamics == "variable") %>%
    select(genome) %>%
    unique() %>% pull()

predY_contrasting <- predY_summ %>%
    pivot_wider(names_from = c(day,treatment), values_from = c(mean,q0.1,q0.9))  %>%
    mutate(TG1 = ifelse(q0.9_7_TG1 < q0.1_35_TG1, "asc", "inv"))  %>%
    mutate(TG1 = ifelse(q0.1_7_TG1 > q0.9_35_TG1, "dec", "inv"))  %>%
    mutate(TG2 = ifelse(q0.9_7_TG2 < q0.1_35_TG2, "asc", "inv"))  %>%
    mutate(TG2 = ifelse(q0.1_7_TG2 > q0.9_35_TG2, "dec", "inv"))  %>%
    mutate(TG3 = ifelse(q0.9_7_TG3 < q0.1_35_TG3, "asc", "inv"))  %>%
    mutate(TG3 = ifelse(q0.1_7_TG3 > q0.9_35_TG3, "dec", "inv"))  %>%
    mutate(TG4 = ifelse(q0.9_7_TG4 < q0.1_35_TG4, "asc", "inv"))  %>%
    mutate(TG4 = ifelse(q0.1_7_TG4 > q0.9_35_TG4, "dec", "inv"))  %>%
    mutate(TG5 = ifelse(q0.9_7_TG5 < q0.1_35_TG5, "asc", "inv"))  %>%
    mutate(TG5 = ifelse(q0.1_7_TG5 > q0.9_35_TG5, "dec", "inv"))  %>%
    select(TG1,TG2, TG3, TG4, TG5) %>%
    mutate(comparison = ifelse(any(c(TG1, TG2, TG3, TG4, TG5) != TG1), "contrasting", "aligned")) %>%
    filter(comparison == "contrasting") %>%
    select(genome) %>%
    unique() %>% pull()

```

```{r hmsc_temporal_predictions_genomes_plot}
predY %>%
  filter(!is.infinite(value) | is.na(value)) %>%
  filter(genome %in% predY_contrasting)  %>% #only display genomes with contrasting dynamics across treatments
  group_by(genome, day, treatment) %>%
  summarize(value = mean(value, na.rm = TRUE)) %>%
  left_join(genome_metadata, by=join_by(genome == genome)) %>%
  ggplot(aes(x=day,y=value, group=genome, color=genome)) + 
      geom_line() +
      facet_grid(genus ~ treatment) +
      labs(y="Genome logfold abundance",x="Day") +
      theme(legend.position = "none",
            strip.text.y = element_text(angle = 0))


```

```{r hmsc_temporal_predictions_traits, warning=FALSE, comments="", message=FALSE}

#Treatment-specific trait predictions
predT_TG1 <- lapply(pred_TG1, function(a) (a %*% m$Tr)/matrix(rep(rowSums(a), m$nt), ncol = m$nt)) %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG1",gradientlength*1000)) %>%
        select(-'(Intercept)') %>%
        pivot_longer(-c(day,treatment), names_to = "trait", values_to = "value")

predT_TG2 <- lapply(pred_TG2, function(a) (a %*% m$Tr)/matrix(rep(rowSums(a), m$nt), ncol = m$nt)) %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG2",gradientlength*1000)) %>%
        select(-'(Intercept)') %>%
        pivot_longer(-c(day,treatment), names_to = "trait", values_to = "value")

predT_TG3 <- lapply(pred_TG3, function(a) (a %*% m$Tr)/matrix(rep(rowSums(a), m$nt), ncol = m$nt)) %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG3",gradientlength*1000)) %>%
        select(-'(Intercept)') %>%
        pivot_longer(-c(day,treatment), names_to = "trait", values_to = "value")

predT_TG4 <- lapply(pred_TG4, function(a) (a %*% m$Tr)/matrix(rep(rowSums(a), m$nt), ncol = m$nt)) %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG4",gradientlength*1000)) %>%
        select(-'(Intercept)') %>%
        pivot_longer(-c(day,treatment), names_to = "trait", values_to = "value")

predT_TG5 <- lapply(pred_TG5, function(a) (a %*% m$Tr)/matrix(rep(rowSums(a), m$nt), ncol = m$nt)) %>%
        do.call(rbind,.) %>%
        as.data.frame() %>%
        mutate(day=rep(gradient,1000)) %>%
        mutate(treatment=rep("TG5",gradientlength*1000)) %>%
        select(-'(Intercept)') %>%
        pivot_longer(-c(day,treatment), names_to = "trait", values_to = "value")

predT <- rbind(predT_TG1,predT_TG2,predT_TG3,predT_TG4,predT_TG5)
```

```{r hmsc_temporal_predictions_traits_plot}
predT %>%
  filter(day == 7 | day == 14 | day == 21 | day == 28 | day == 35) %>%
  filter(value < 1) %>%
  filter(value > 0) %>%
  mutate(day = factor(day)) %>%
  ggplot(aes(x=value,y=fct_rev(day))) + 
      geom_boxplot(outlier.size=-1) +
      facet_grid(treatment ~ trait) +
      labs(y="Day",x="Metabolic capacity index")
```
